### DISCUSSION TITLE
How does make_rng() work in Flax/JAX?

### DISCUSSION THREAD

**Discussion by marcvanzee** [google/flax#1998] [Q&A] [answered]
(Original question from @lucasb-eyer, rephrased by me)

How does Flax Linen's [Module.make_rng()](https://flax.readthedocs.io/en/latest/flax.linen.html?highlight=make_rng#flax.linen.Module.make_rng) work?

**Comment by marcvanzee** [comment, marked as answer]
(Original answer by @levskaya, rephrased by me):

When you call [Module.init()](https://flax.readthedocs.io/en/latest/flax.linen.html?highlight=make_rng#flax.linen.Module.init) or [Module.apply()](https://flax.readthedocs.io/en/latest/flax.linen.html?highlight=make_rng#flax.linen.Module.apply), you pass them a set of random keys, which are treated as "root" keys (note that init allows a single key as shorthand for `{'params': rng}`). At each submodule boundary we fold-in a hash of the submodule name for each of these root keys, and within a submodule where `self.make_rng('foo')` is called we keep track of a counter that's also folded in to guarantee uniqueness of the key with each call.

**Comment by wbrenton** [comment]
is there a good example of make_rng being used?

**Reply by marcvanzee** [reply]
You could take a look at the [Dropout Module](https://github.com/google/flax/blob/842a28748999d03b042463c3e38b2a77dc54d85f/flax/linen/stochastic.py#L81)

**Reply by wbrenton** [reply]
Thanks!
